### PR TITLE
Automate upgrade using `moved` blocks

### LINKED ISSUE #63: Support nullable flag for variables with default values
## Describe the Feature

Add nullable flag (`nullable = false`) for module input variables which have a default value and where `null` is not a sensible/handled value for the variable.

This should not be a breaking change since, as far as I can tell, the module does not currently handle null values for variables.

I am happy to tackle this feature request if you are happy with it.

## Expected Behavior

Module input variables (satisfying the description, above) use the nullable flag and set the value to `false` which makes them assume their default value when a `null` input value is supplied.

## Use Case

This is useful when consuming the module generically (e.g. using Terragrunt for different environments):

```hcl
terraform {
  source = "${local.base_source_url}?ref=vx.x.x"
}

locals {
  # Automatically load variables
  account_vars = read_terragrunt_config(find_in_parent_folders("account.hcl"))
  service_vars = read_terragrunt_config(find_in_parent_folders("service.hcl"))

  # Extract out common variables for reuse
  account_name = local.account_vars.locals.account_name

  base_source_url = "<some-source-url>"
}


# ---------------------------------------------------------------------------------------------------------------------
# MODULE PARAMETERS
# These are the variables we have to pass in to use the module. This defines the parameters that are common across all
# environments.
# ---------------------------------------------------------------------------------------------------------------------  
inputs = {
  namespace = "${local.account_name}"
  stage = local.service_vars.locals.s3_log_stage
  name = local.service_vars.locals.s3_log_name
  acl = local.service_vars.locals.s3_log_acl
  standard_transition_day = local.service_vars.locals.s3_log_standard_transition_day
  glacier_transition_days = local.service_vars.locals.s3_log_glacier_transition_days
  expiration_days = local.service_vars.locals.s3_log_expiration_days
  allow_encrypted_uploads_only = local.service_vars.locals.s3_log_allow_encrypted_uploads_only
  allow_ssl_requests_only = local.service_vars.locals.s3_log_allow_ssl_requests_only
  tags = local.service_vars.locals.s3_log_tags
  policy = local.service_vars.locals.s3_log_policy
}
```

By making the input variables nullable, we can set ours to `null` and get the default values without having any knowledge of the default values. We can also easily override the default values on other environments, where needed. 

## Describe Ideal Solution

Add nullable flag (`nullable = false`) for module input variables which have a default value and where `null` is not a sensible/handled value for the variable. This will include a Terraform required version >= 1.1.

## Additional Context

When calling a module and supplying `null` for one of the input parameters, Terraform's default behaviour was always to set the actual value of the variable to `null`. There is a thread where many people have contested this behaviour: https://github.com/hashicorp/terraform/issues/24142

Terraform 1.1 has released a `nullable` flag which allows module creators to choose whether null is a valid value for an input variable, or whether the variable should take on the default value when set to `null`: https://github.com/hashicorp/terraform/pull/29832

### PR DESCRIPTION
**Pull request by Nuru**
## Important Notes
 
- Terraform version 1.3.0 and Terraform AWS version 4.9.0 or later are required
- You can now select default values for (non-deprecated) inputs by setting them to `null`
- With Terraform 1.3 the manual interventions documented for upgrading to this module's versions 0.27.0 and 0.28.0 are no longer needed. You can safely upgrade from any earlier version to this one (although we always recommend leaving `force_destroy` at its default value of `false`, and if you have it set to `true` but want extra safety against the S3 bucket being destroyed, set it to `false` _**before**_ upgrading).
- The `force_destroy_enabled` flag introduced in v0.27.0 has been removed
- In version 0.28.0, old lifecycle rule variables were deprecated and the new `lifecycle_configuration_rules` input was introduced. In that version, you would continue to get the old default lifecycle rule even if you supplied new rules via `lifecycle_configuration_rules`. Now, the default behavior is to ignore all the deprecated lifecycle inputs when the `lifecycle_configuration_rules` input is not empty, unless you explicitly set `lifecycle_rule_enabled` to true. 

## what

- Automate the upgrade process from v0.26.0 or earlier by using `moved` block functionality introduced in Terraform 1.3.0
-  Add `nullable = false` for module input variables which have a default value and where null is not a sensible/handled value for the variable.

## why

- Safely upgrade without loss of data or manual intervention
- Allow users to select default values by setting inputs to `null`, closes #63

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -20,32 +20,13 @@ moved {
   from = aws_s3_bucket_public_access_block.default
   to   = module.aws_s3_bucket.aws_s3_bucket_public_access_block.default
 }
-*/
+moved {
```
Comment (nitrocode) [inline main.tf:23]: It would be nice to move these moved blocks to a separate file. Perhaps a moved.tf?

**Review by nitrocode — DISMISSED** [review]

**Review by nitrocode — APPROVED** [review]

File: variables-deprecated.tf [hunk 2]
```diff
@@ -1,101 +1,103 @@
...
   description = "(Deprecated, use `lifecycle_configuration_rules` instead)\nSpecifies when non-current object versions expire (in days)"
+  default     = null
 }
 
 
 
 locals {
   deprecated_lifecycle_rule = {
-    enabled = var.lifecycle_rule_enabled
+    enabled = var.lifecycle_rule_enabled == true || (var.lifecycle_rule_enabled == null && length(var.lifecycle_configuration_rules) == 0)
```
Comment (nitrocode) [inline variables-deprecated.tf:82]:
```suggestion
    enabled = var.lifecycle_rule_enabled || (var.lifecycle_rule_enabled == null && length(var.lifecycle_configuration_rules) == 0)
```

**Review by nitrocode — APPROVED** [review]

File: variables-deprecated.tf (hunk 2 above)
Comment (Nuru) [reply variables-deprecated.tf:82]: @nitrocode 
```
│ Error: Operation failed
│ 
│   on variables-deprecated.tf line 82, in locals:
│    82:   enabled = var.lifecycle_rule_enabled || (var.lifecycle_rule_enabled == null && length(var.lifecycle_configuration_rules) == 0)
│      ├────────────────
│      │ var.lifecycle_rule_enabled is null
│ 
│ Error during operation: argument must not be null.
```

Comment (nitrocode) [reply variables-deprecated.tf:82]: ah interesting, I didn't realize if a var `type = bool` that defaulted to `null` wouldn't already implicitly resolve to `false`. Good to know.

**Pull request merged**